vowpalwabbit/binary.cc

```diff
@@ -4,44 +4,62 @@
 
 #include <cfloat>
 #include "reductions.h"
+#include "debug_log.h"
 
 using namespace VW::config;
+using std::endl;
 
-template <bool is_learn>
-void predict_or_learn(char&, VW::LEARNER::single_learner& base, example& ec)
-{
-  if (is_learn)
-    base.learn(ec);
-  else
-    base.predict(ec);
+VW_DEBUG_ENABLE(false)
 
-  if (ec.pred.scalar > 0)
-    ec.pred.scalar = 1;
-  else
-    ec.pred.scalar = -1;
-
-  if (ec.l.simple.label != FLT_MAX)
+namespace VW { namespace binary {
+  template <bool is_learn>
+  void predict_or_learn(char&, VW::LEARNER::single_learner& base, example& ec)
   {
-    if (fabs(ec.l.simple.label) != 1.f)
-      std::cout << "You are using label " << ec.l.simple.label << " not -1 or 1 as loss function expects!" << std::endl;
-    else if (ec.l.simple.label == ec.pred.scalar)
-      ec.loss = 0.;
+    if (is_learn)
+    {
+      VW_DBG(ec) << "binary: before-base.learn() " << simple_label_to_string(ec) << features_to_string(ec) << endl;
```

Comment:
will the debug statements remain? shouldn't they be ad-hoc?

Reply:
VW_DBG is designed to be compiled away in release mode.  This should not have a runtime impact.
VW_DEBUG_ENABLE can be use to enable/disable VW_DBG in individual reductions for release mode tracing of vw execution.

This is used for example to track down predict/learn issues with continuous actions.  I am inclined to leave these in for now.  Perhaps this can be revisited if these statements become too prolific.

I was also thinking about enhancing VW_DBG to write to some other endpoint to enable 

> Self describing means that a parser could read the file and know what label type it contains without the user needing to specify it. This is how CCB and Slates work, for example: `ccb shared | ...`
> 
> This is something that John has mentioned we should move towards, you can check the meeting chat for John's comment.

I don't see it..  VW meeting chat?  

I am not sure this was fully discussed.  There are parsers today in vw that populate different labels types using the same format.  We would loose the ability to reinterpret the data as a different label type with this approach.  Maybe this is ok, maybe not.   I think we should discuss this further away from this PR.

Reply:
I don't see much value in including in the source code custom debug calls that get called only during debug mode. Not to mention, all this functionality can be replicated today, without re-compilation using the custom python reductions. So I'd be wary of this sort of approach that litters the code with custom debug lines that will change depending on what you are investigating.

Reply:
I agree with @lalo, debugging code like this doesn't seem like something we want to check into master, but is rather used during an investigation.

Reply:
When working with many nested reductions and multiple learners in the tree, I realized that VW lacks good execution trace facility common in other production systems.  Introducing breakpoints and adding ad-hoc debug statements might still be necessary but being able to see predictions returned by nested learner in context of the stack helped with quick diagnosis of data and learning issues.

This not entirely without precedent in vw code.  See search reduction for example.  VW_DBG is a generalization of that.  I was going to replace the mechanism in search but did not want to do that in this PR.

The AutoML type reductions (which will go on top of continuous actions) will benefit from kind of execution tracing .

Reply:
Rajan, I understand the void you are trying to fill but I don't see this statement as fulfilling a tracing/log system. A more interesting approach would be to create a Tracing reduction that gets braided in optionally between all reductions without them even being aware of it (via setup_base). If you notice, that debug print statement you added does not refer to anything in that file just `example` related facilities. Therefore, basing it on the current PyAtom work you can have a cpp class that looks something like this (this is just a brief sketch, there's just one extra consideration to have around the specific types but nothing that templates can't fix I guess):

```
class ExampleTracer : ExternalBinding
{
            ExampleTracer(std::string m_NameOfNexReduction) {}
            void SetBaseLearner(void* learner) { // setting learner };
            void ActualLearn(example *) { 
                  print<< "{$m_NameOfNexReduction}: before-base.learn() " << simple_label_to_string(ec) << features_to_string(ec) << endl;
                  base.predict();
                  print<< "{$m_NameOfNexReduction}: after-base.learn() " << simple_label_to_string(ec) << features_to_string(ec) << endl;
            }
            void ActualPredict(example *) { 
                  print<< "{$m_NameOfNexReduction}: before-base.predict() " << scalar_pred_to_string(ec) << features_to_string(ec) << endl;
                  base.predict();
                  print<< "{$m_NameOfNexReduction}: after-base.predict() " << scalar_pred_to_string(ec) << features_to_string(ec) << endl;
            }
            bool ShouldRegisterFinishExample() { return false; }
            void ActualFinishExample(example *) { }
}
```

Then you push one of these at the top of the stack calling it with ExampleTracer("Begin") and then on setup_base every time a setup function returns into setup_base you pop another ExampleTracer(current_name); and with this you would achieve a tracer that is printing the mutation between reductions in examples.

The benefits of something like this approach would be 1) not to have to recompile 2) flexibility on anyone could create a class that extends those functionality and print whatever they want 3) devs could still use your VW_DBG for ad-hoc debugging 4) it's agnostic to reductions, since it's only caring about examples

```
Added Reduction: Tracer("gd")
Added Reduction: gd
Added Reduction: Tracer("scorer-identity")
Added Reduction: scorer-identity
Added Reduction: Tracer("csoaa")
Added Reduction: csoaa
Added Reduction: Tracer("search")
Added Reduction: search
Added Reduction: Tracer("begin")
```

I haven't coded this, but it's an interesting concept to think about. Anyways, if you want tracing why not augment it with verbosity levels, leverage the current design of VW, do proper tracing ids, evolve `audit`, etc. Anyways, it's an interesting problem (tracing) that deserves at least some discussion instead of a comment in an unrelated PR.